src/rust/catloop/futures/connect.rs

```diff
@@ -46,288 +47,192 @@ use ::std::{
 /// This was chosen arbitrarily.
 const MAX_ACK_RECEIVED_ATTEMPTS: usize = 1024;
 
-//======================================================================================================================
-// Enumerations
-//======================================================================================================================
-
-/// Client-side states in the connection establishment protocol.
-enum ClientState {
-    InitiateConnectRequest {
-        qt_rx: Option<QToken>,
-    },
-    ConnectRequestSent {
-        qt_tx: QToken,
-        qt_rx: Option<QToken>,
-    },
-    ConnectAckReceived {
-        attempt: usize,
-        qt_rx: QToken,
-    },
-    Connected {
-        qt_tx: QToken,
-        remote: SocketAddrV4,
-        duplex_pipe: Rc<DuplexPipe>,
-    },
+pub async fn connect_coroutine(
+    catmem: Rc<RefCell<CatmemLibOS>>,
+    remote: SocketAddrV4,
+    yielder: Yielder,
+) -> Result<(SocketAddrV4, Rc<DuplexPipe>), Fail> {
+    let ipv4: &Ipv4Addr = remote.ip();
+    let port: u16 = remote.port().into();
+
+    // Open connection to server.
+    let control_duplex_pipe: Rc<DuplexPipe> = Rc::new(DuplexPipe::open_duplex_pipe(catmem.clone(), ipv4, port)?);
+
+    // Gets the port for the new connection from the server by sending a connection request repeatedly until a port
+    // comes back.
+    let new_port: u16 = get_port(catmem.clone(), control_duplex_pipe.clone(), &yielder).await?;
```

Comment:
If we fail, won't we leave behind a dangling control duplex pipe opened?

Reply:
Yes,  thanks for catching that!

Reply:
Thanks!